tensorflow/java/src/gen/cc/source_writer.cc

```diff
@@ -47,16 +75,265 @@ SourceWriter& SourceWriter::Write(const string& str) {
   return *this;
 }
 
+SourceWriter& SourceWriter::Append(const StringPiece& str) {
+  if (!str.empty()) {
+    if (newline_) {
+      DoAppend(left_margin_ + line_prefix_);
+      newline_ = false;
+    }
+    DoAppend(str);
+  }
+  return *this;
+}
+
+// Writes the signature of a Java type.
+SourceWriter& SourceWriter::Append(const Type& type) {
+  if (type.kind() == Type::Kind::GENERIC && type.name().empty()) {
+    Append("?");
+  } else {
+    Append(type.name());
+  }
+  if (!type.parameters().empty()) {
+    Append("<");
+    for (std::vector<Type>::const_iterator it = type.parameters().cbegin();
```

Comment:
Can we use range iteration here and in other places instead?

```c++
bool first = true;
for (const Type& t : type.parameters()) {
  if (!first) {
    Append(", ");
  }
  first = false;
  Append(t);
}  
```

Reply:
of course...